packages/lodestar/src/chain/chain.ts

```diff
@@ -185,6 +186,30 @@ export class BeaconChain extends (EventEmitter as { new(): ChainEventEmitter })
     };
   }
 
+  /**
+   * Restore forkchoice state from database.
+   * TODO: if performance is not good, we may store the whole forkchoice to db on stop()
+   * and compare that against last block in db.
+   */
+  public async restoreForkChoice(): Promise<void> {
+    const start = Date.now();
+    const blocks = await this.db.block.getAll() || [];
+    for (const block of blocks) {
+      const stateRoot = block.message.stateRoot.valueOf() as Uint8Array;
+      const state = await this.db.state.get(stateRoot);
+      await this.forkChoice.addBlock({
+        slot: block.message.slot,
+        blockRootBuf: this.config.types.BeaconBlock.hashTreeRoot(block.message),
+        stateRootBuf: stateRoot,
+        parentRootBuf: block.message.parentRoot.valueOf() as Uint8Array,
+        justifiedCheckpoint: state.currentJustifiedCheckpoint,
+        finalizedCheckpoint: state.finalizedCheckpoint
+      });
+      await this.attestationProcessor.receiveBlock(block);
+    }
+    this.logger.info(`Restore forkchoice in ${Math.floor(Date.now() - start)/1000} seconds, numblock=${blocks.length}`);
```

Comment:
One possibility is to store the whole forkchoice to db and do simple check against latest db block to use it. If the check returns false we will build forkchoice from scratch like this.

Reply:
@wemeetagain I don't see Record/Map as basic ssz type so we can't store node list as ssz serialize(), can you confirm? The option for me is to use `JSON.stringnify()` to serialize node list and store to db (thanks to protoarray impl), when we start the chain we'll restore forkchoice and compare its head root against with `db.chain.getChainHeadRoot()`. If it's good we'll use it, otherwise we have to build forkchoice manually. Wdyt?

The other concern is how to restore AttestationAggregator and `bestJustifiedCheckpoint` inside the store. For this, I'd load all blocks and call `addBlock(), addAttestation()` manually to make sure it's good (as long as we don't load states, I think it's ok to load all blocks). Do you think it's necessary, is it heavy to load all blocks and process one by one?

Reply:
other option is probably to write `ForkChoiceNode` lodestar type with `childrenHexes: string[], childrenIndexes: number[]` to address the `children` Record/map inside the node and use our ssz serialize. Then we need to convert `ForkChoiceNode` and the `Node` type inside forkchoice store.